node/src/components/block_proposer.rs

```diff
@@ -364,263 +385,62 @@ impl BlockProposerReady {
     }
 
     /// Returns a list of candidates for inclusion into a block.
-    // TODO:
-    /// rename to proposed deploys
-    // TODO:
-    /// maybe use cuckoofilter
-    fn propose_deploys(
+    fn propose_proto_block(
         &mut self,
         deploy_config: DeployConfig,
         block_timestamp: Timestamp,
-        past_deploys: HashSet<DeployHash>,
-    ) -> HashSet<DeployHash> {
-        // deploys_to_return = all deploys in pending that aren't in finalized blocks or
+        past_deploys: &[DeployHash],
+        random_bit: bool,
+    ) -> ProtoBlock {
+        // all deploys in pending that aren't in finalized blocks or
         // proposed blocks from the set `past_blocks`
-        self.sets
+        let deploys = self
+            .sets
             .pending
             .iter()
-            .filter(|&(hash, deploy)| {
-                self.is_deploy_valid(deploy, block_timestamp, &deploy_config, &past_deploys)
-                    && !past_deploys.contains(hash)
-                    && !self.sets.finalized_deploys.contains_key(hash)
+            .filter_map(|(hash, deploy)| {
+                if let DeployType::Deploy(header) = deploy {
+                    if self.is_deploy_valid(header, block_timestamp, &deploy_config, past_deploys)
+                        && !past_deploys.contains(hash)
+                        && !self.sets.finalized_deploys.contains_key(hash)
+                    {
+                        Some(*hash)
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
             })
-            .map(|(hash, _deploy)| *hash)
             .take(deploy_config.block_max_deploy_count as usize)
-            .collect::<HashSet<_>>()
-        // TODO: check gas and block size limits
+            .collect::<Vec<_>>();
+
+        let transfers = self
+            .sets
+            .pending
+            .iter()
+            .filter_map(|(hash, deploy)| {
+                if let DeployType::Transfer(header) = deploy {
+                    if self.is_deploy_valid(header, block_timestamp, &deploy_config, past_deploys)
+                        && !past_deploys.contains(hash)
+                        && !self.sets.finalized_deploys.contains_key(hash)
+                    {
+                        Some(*hash)
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
```

Comment:
This could be deduplicated if it weren't for the transfer vs. deploy distinction. Maybe `self.pending` should be split into two collections anyway?

Reply:
I resolved the duplication here using some iterators and `fold`.

Reply:
Yes, but you now first separate them, then chain them again, and in the closure separate them again; that doesn't make sense to me.
Also: You _first_ limit the transfer/deploy count and _then_ filter. That's the wrong way around, and will usually result in much fewer transfers and deploys than allowed.

Why not extract the filter (the "valid and not a past/finalized deploy" part) into a closure or method and apply it to both iterators, and _after_ filtering limit the number (`take(…_count as usize)`)?

Reply:
> Maybe `self.pending` should be split into two collections anyway?

I agree with Andreas here – we might end up iterating over the same elements twice. I.e. we will be iterating over `transfer` when picking `deploys` and vice versa. This is suboptimal performance-wise.